### ISSUE TITLE
When using the SentenceTransformer.encode() method, the CPU usage reaches 100%

### ISSUE THREAD

**Issue by hello1534** [UKPLab/sentence-transformers#3091]
When using HuggingFaceEmbeddings in LangChain to embed documents, I noticed that the embedding process takes significantly longer on the server compared to my local machine. My local computer has only an 8-core CPU, while the server has more than 90 cores. Logically, the server's CPU performance should be better, and the process should be faster. Upon checking the code, I found that the SentenceTransformer.encode method is being used for embedding.

So I have two questions: First, compared to encode_multi_process, the encode method appears to be single-threaded, but why does CPU usage on both my local machine and the server remain almost 100% during usage? Second, can I use the encode_multi_process method to improve the embedding speed on a multi-core server, or are there any recommended ways to enhance embedding speed?

model is shibing624/text2vec-base-chinese 
other param
---- {'device': None} --- {'normalize_embeddings': True} -----
SentenceTransformer(
  (0): Transformer({'max_seq_length': 128, 'do_lower_case': False}) with Transformer model: BertModel 
  (1): Pooling({'word_embedding_dimension': 768, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': False, 'pooling_mode_lasttoken': False, 'include_prompt': True})
)

**Comment by tomaarsen** [comment]
Hello!

It's indeed possible that your server (despite having more cores) is weaker when single-threaded. And indeed, `encode` does not use multiple processes. 100% CPU usage is not strictly a bad thing either, this would actually be indicative of good usage of the available hardware I imagine.

But yes, on many CPU-only devices it's possible to speed up inference/embedding speed with `encode_multi_process`. I don't have great advice when it comes to parameters, as it all depends on the hardware. My primary recommendation here is to load a bunch of texts and benchmark the inference time using those:
```python
from datasets import load_dataset

dataset = load_dataset("sentence-transformers/law-gpt", subset="triplet", split="train")
texts = dataset["anchor"] # <- 500 texts
```

And yes, there is (new) documentation on how to speed up inference: https://sbert.net/docs/sentence_transformer/usage/efficiency.html
Depending on your hardware, you might get very significant speedups with OpenVINO int8 quantization.
![backends_benchmark_cpu](https://github.com/user-attachments/assets/ff208c9b-836e-497a-97ff-340cf56a0725)

I actually did tests on https://huggingface.co/shibing624/text2vec-base-chinese specifically. See https://huggingface.co/shibing624/text2vec-base-chinese/discussions/30#6734ca2d0470f450e64e1d69 for details.
In short, OpenVINO static int8 quantization with Chinese STSB resulted in a performance decrease of 3.6% on some datasets to a performance increase of 1.88% on others. It should also allow for about a 4x speedup. [This documentation](https://sbert.net/docs/sentence_transformer/usage/efficiency.html#quantizing-openvino-models) should explain how to quantize with OpenVINO.

Otherwise, you can use "normal" OpenVINO or ONNX for slight improvements (again: do benchmark. For example, the performance improvements can be much better for shorter texts)

Something to note is that if you load a model with a non-torch backend, then it's possible that it's not also compatible with `encode_multi_process`.

- Tom Aarsen

**Comment by hello1534** [comment]
Thank you for your reply. I still don't quite understand why the CPU usage can reach 100% without using multiprocessing in encode. Doesn't Python's GIL  limit the CPU usage? Or is it due to PyTorch's characteristics? Also, if the CPU usage of encode is already at 100%, would using encode_multi_process not bring any performance improvement?

**Comment by HenningDinero** [comment]
>Doesn't Python's GIL limit the CPU usage?

GIL does not limit the CPU usage but (losely speaking) limits the interpreter to only run 1 python-code at a time i.e if you are multi-threading each thread has to wait for the GIL to release before it can execute some python-code.

**Comment by Jakobhenningjensen** [comment]
>Doesn't Python's GIL limit the CPU usage?

GIL does not limit the CPU usage but (losely speaking) limits the interpreter to only run 1 python-code at a time i.e if you are multi-threading each thread has to wait for the GIL to release before it can execute some python-code thus the threads are waiting for their turn to execute some code (again, losely speaking)
